File: LeetCode75-sols/Array_String/JavaScript/Greatest_Common_Divisor_of_Strings.js
/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */
var gcdOfStrings = function(str1, str2) {
    if (str1 + str2 !== str2 + str1) {
      return "";
    }
  // Compute GCD of lengths
  const gcdLength = gcd(str1.length, str2.length);

  // Return substring from 0 to gcdLength
  return str1.substring(0, gcdLength);
};

// Helper function fo rGCD
function gcd(a, b) {
  if (b == 0) {
    return a;
  }
  return gcd(b, a % b);

File: LeetCode75-sols/Array_String/Rust/Greatest_Common_Divisor_of_Strings.rs
impl Solution {
    pub fn gcd_of_strings(str1: String, str2: String) -> String {
        // Check if concatenation of str1+str2 = str2+str1
        if (str1.clone() + &str2 != str2.clone() + &str1) {
            return String::new();
        }

        let gcd_length = gcd(str1.len(), str2.len());

        // Return substring
        str1[..gcd_length].to_string()
    }
}

// Helper GCD function
fn gcd(a: usize, b: usize) -> usize {
    if (b == 0) {
        a
    } else {
        gcd (b, a % b)
    }
}
